### ISSUE TITLE
[FEATURE] Connect to localhost database as default

### ISSUE THREAD

**Issue by creativiii** [sprintcube/docker-compose-lamp#122]
**About which PHP Version we are talking ?**
Any
**Is your feature request related to a problem? Please describe.**
In most LAMP server environments it's unlikely an application's hostname will be anything other than `localhost`. Unfortunately the way the docker compose is set up this doesn't seem possible. While it's possible to change default hostname to anything other than `database`, setting it to localhost will throw an error.

I tried modifying the `docker-compose.yml` [myself](https://pastebin.com/zgT4JZSd) but I haven't been able to get it working as I wanted.

**Describe the solution you'd like**
Instead of having 'database' as the default hostname for mysql, have `localhost` since it seems harder to accomplish. 

I'll take even just a simple addition to the readme explaining if this can be done and how to do it.

**Comment by Egahras** [comment]
What did you change in the docker-compose.yml to get what behavior?

If I understand you correct, you want the webserver service be able to connect the database with localhost? I think the only possible way is to set up a dns rule in the webserver service that will point requests on localhost:3306 to database:3306.

I don't believe a single setting in docker-compose can achieve that.
First I thought of:

```
extra_hosts:
  - localhost:${COMPANY_DB_IP}
```
But if I understand this setting correctly, this would result in each call of localhost, no matter what port, will point you to to the defined ip!?

**Comment by creativiii** [comment]
> What did you change in the docker-compose.yml to get what behavior?

After doing some research and asking on Stackoverflow someone mentioned I could add `network_mode: host` to the webserver container would allow me to do that. I then also had to remove the `link` because they didn't work with the `network_mode` setting.

Of course it didn't work.

It just seems weird that this isn't possible? It's such a common thing in most servers, I find it hard to believe I'd have to use a workaround to set it up.

**Comment by Egahras** [comment]
It is possible.
You could create your own image and install both, the webserver and the databae within 1 image/system/container.
I see that the name of this docker-compose - package is not quite accurate. LAMP means in common case a single machine with Linux as operating system and Apache,MySQL and PHP as provided services. But It's as common that these things are seperated to provide scaleability and reliability/redundancy.
And this scaleability + Flexibility are the reason why we want to use docker. 
This package is meant as dev-environment. So you can easily switch the php version, or try another mysql version/server such as mariadb.

Most Application I know provide settings/configs where you can define the host of the mysql server.

Maybe there is a magical function/Setting in the docker universe that can provide the magic youre searching for. But as far as i know. Localhost is "untouchable" due to the fact its mapped to 127.0.0.1 or (ipv6) ::1

So you can try to add /configure a dns setup in the webserver container which reroutes all localhost:3306 calls to the database host. But please make it configurable via an environment variable, so everybody can decide if they want to use it or not.

Or you can just update the projects you want use, so that their databes-hostname is configurable.

**Comment by MrOffline77** [comment]
Hi, much text here thats cool 👍 

@creativiii I definetly get your point with the `localhost`!= `database`thing. 
The historic reason for this is, that in the vertical scaled world ( e.g. shared webhosting) everything runs most of the time on one linux box. But there are also a lot of companies using a separated database cluster to provide solid database performance. 

Docker in general creates a separated bridged network for all docker containers, this network can be connected to your Public site of things or to your localhost loopback device. 
Docker-Compose uses the same technique, docker-compose or docker is aware of the other containers inside this capsulated network. Therefore ist possible to communicate with other containers inside docker-compose with their descriptive names, defined in the docker-compose.yaml. To change this behavior ins this case, you need to touch quite a lot of things and build your way around it. Maybe it's possible, but the code will get completely unorganized and gain a lot complexity. Btw, same thing for redis and so on. 

So in my opinion we should give this a place in the readme, maybe with some demo `.env` files in the repo from Laravel or symphony, preconfigured with the connections. This should make it easier.

@Egahras 
`You could create your own image and install both, the webserver and the databae within 1 image/system/container.`

You could yeah, but this is definitely not best practice. 1 Container = 1 running Process. Docker only checks PID-1, if PID-2 dies, nothing happen and your container is broken, and no one will ever notice. If PID-1 dies, docker restarts your container, no outage no unexpected behavior. Restore to defined state. 

@Egahras 
`Most Application I know provide settings/configs where you can define the host of the mysql server.`
yes exactly almost every Framework or PHP `$thing` is providing some kind of dot env file. 

You could also create some sort of dnsmasq'ing, but most of the time this creates more problems than it solves. Even if you refer to localhost, your localhost is not like `localhost`cause docker is creating a own localhost loopback with some special things to it. 

tldr: 
* Feel free to create a pull request with a readme change, where you can explain the behavior. 
* The part with the demo dot envs I'm not quite sure about, I generally assume that a developer is able to change a name in a config file, that shouldn't be that hard. But I'm looking from a different perspective. Maybe someone can convince me to accept a pull req, with a demo directory and some demo dot envs. But at the moment I'm not sure if there is a real world use case for this. Free feel to argue about this :).

**Comment by creativiii** [comment]
My only real world use case is not as strong honestly. I have to work with some _not so great_ CMS that simply don't allow to easily change config to change `localhost`. 

In my head having a docker with LAMP is especially useful for local development, meaning you'd probably want a minimal setup solution where you can just drop your old LAMP server files and have them work without any extra setup.

It's not a very strong argument. I also have no idea how to accomplish this so my points are moot 😂

**Comment by MrOffline77** [comment]
Most of the time the connection settings are somewhere hardcoded in a central file. Just grep the code for `localhost` and modify the value. Maybe this works. Localhost in docker context always means the container itself.

**Comment by MrOffline77** [comment]
Since this is not a code issue on our site, im going to close this issue.
